LibreSSL Support

LibreSSL is a compatible fork of OpenSSL which you can use to compile Telegram Desktop with.
It is a fork by OpenBSD project, with goal of refactoring the code .It was created as a response to the Heartbleed vulnerability.
To this date, it has had less vulnerabilities compared to OpenSSL, because it has better code audit than OpenSSL.

Please provide proper compile instructions for using LibreSSL with Telegram Desktop.

As of tdesktop 1.2.1 this patch doesn't work anymore ( with libressl 2.6.x):
```/var/tmp/paludis/build/net-im-telegram-desktop-1.2.1/work/telegram-desktop-1.2.1/Telegram/SourceFiles/mtproto/rsa_public_key.cpp: In constructor 'MTP::internal::RSAPublicKey::Private::Private(base::const_byte_span, base::const_byte_span)':
/var/tmp/paludis/build/net-im-telegram-desktop-1.2.1/work/telegram-desktop-1.2.1/Telegram/SourceFiles/mtproto/rsa_public_key.cpp:119:9: error: 'RSA_set0_key' was not declared in this scope
    if (!RSA_set0_key(_rsa, n, e, nullptr) || !valid) {
         ^~~~~~~~~~~~
/var/tmp/paludis/build/net-im-telegram-desktop-1.2.1/work/telegram-desktop-1.2.1/Telegram/SourceFiles/mtproto/rsa_public_key.cpp:119:9: note: suggested alternative: 'RSA_check_key'
    if (!RSA_set0_key(_rsa, n, e, nullptr) || !valid) {
         ^~~~~~~~~~~~
         RSA_check_key
/var/tmp/paludis/build/net-im-telegram-desktop-1.2.1/work/telegram-desktop-1.2.1/Telegram/SourceFiles/mtproto/rsa_public_key.cpp: In member function 'base::byte_vector MTP::internal::RSAPublicKey::Private::getN() const':
/var/tmp/paludis/build/net-im-telegram-desktop-1.2.1/work/telegram-desktop-1.2.1/Telegram/SourceFiles/mtproto/rsa_public_key.cpp:129:3: error: 'RSA_get0_key' was not declared in this scope
   RSA_get0_key(_rsa, &n, nullptr, nullptr);
   ^~~~~~~~~~~~
/var/tmp/paludis/build/net-im-telegram-desktop-1.2.1/work/telegram-desktop-1.2.1/Telegram/SourceFiles/mtproto/rsa_public_key.cpp:129:3: note: suggested alternative: 'RSA_check_key'
   RSA_get0_key(_rsa, &n, nullptr, nullptr);
   ^~~~~~~~~~~~
   RSA_check_key
/var/tmp/paludis/build/net-im-telegram-desktop-1.2.1/work/telegram-desktop-1.2.1/Telegram/SourceFiles/mtproto/rsa_public_key.cpp: In member function 'base::byte_vector MTP::internal::RSAPublicKey::Private::getE() const':
/var/tmp/paludis/build/net-im-telegram-desktop-1.2.1/work/telegram-desktop-1.2.1/Telegram/SourceFiles/mtproto/rsa_public_key.cpp:135:3: error: 'RSA_get0_key' was not declared in this scope
   RSA_get0_key(_rsa, nullptr, &e, nullptr);
   ^~~~~~~~~~~~
/var/tmp/paludis/build/net-im-telegram-desktop-1.2.1/work/telegram-desktop-1.2.1/Telegram/SourceFiles/mtproto/rsa_public_key.cpp:135:3: note: suggested alternative: 'RSA_check_key'
   RSA_get0_key(_rsa, nullptr, &e, nullptr);
   ^~~~~~~~~~~~
   RSA_check_key
/var/tmp/paludis/build/net-im-telegram-desktop-1.2.1/work/telegram-desktop-1.2.1/Telegram/SourceFiles/mtproto/rsa_public_key.cpp: In member function 'void MTP::internal::RSAPublicKey::Private::computeFingerprint()':
/var/tmp/paludis/build/net-im-telegram-desktop-1.2.1/work/telegram-desktop-1.2.1/Telegram/SourceFiles/mtproto/rsa_public_key.cpp:188:3: error: 'RSA_get0_key' was not declared in this scope
   RSA_get0_key(_rsa, &n, &e, nullptr);
   ^~~~~~~~~~~~
/var/tmp/paludis/build/net-im-telegram-desktop-1.2.1/work/telegram-desktop-1.2.1/Telegram/SourceFiles/mtproto/rsa_public_key.cpp:188:3: note: suggested alternative: 'RSA_check_key'
   RSA_get0_key(_rsa, &n, &e, nullptr);
   ^~~~~~~~~~~~
   RSA_check_key
```

--- a/Telegram/SourceFiles/core/utils.cpp
+++ b/Telegram/SourceFiles/core/utils.cpp
@@ -295,7 +295,9 @@ namespace ThirdParty {
 		av_lockmgr_register(nullptr);
 
 		CRYPTO_cleanup_all_ex_data();
+#ifndef LIBRESSL_VERSION_NUMBER
 		FIPS_mode_set(0);
+#endif
 		ENGINE_cleanup();
 		CONF_modules_unload(1);
 		ERR_remove_state(0);

--- a/Telegram/SourceFiles/mtproto/rsa_public_key.cpp
+++ b/Telegram/SourceFiles/mtproto/rsa_public_key.cpp
@@ -16,7 +16,7 @@ https://github.com/telegramdesktop/tdesktop/blob/master/LEGAL
 namespace MTP {
 namespace internal {
 namespace {
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 
 // This is a key setter for compatibility with OpenSSL 1.0
 int RSA_set0_key(RSA *r, BIGNUM *n, BIGNUM *e, BIGNUM *d) {

